Krylov solvers

jamtrott opened the pull request:
This pull request changes the default linear solvers to use preconditioned conjugate gradient algorithm with an AMG preconditioner from Hypre.

Before merging, it may be desirable to add some more error handling or a fallback to another preconditioner, just in case Hypre is not supported. (I believe it is optional to include Hypre when compiling PETSc.) Ideally, it would also be nice if the user could specify the tolerances for the solver and maybe other parameters too, instead of having them hard-coded like this.

jamtrott commented on ldrb/ldrb.py:
```diff
@@ -357,18 +358,35 @@ def apex_to_base(
+    ksp_atol = 1e-15
+    ksp_max_it = 10000
+    ksp_error_if_not_converged = False
+    ksp_monitor = True
+    ksp_view = True
+    pc_type = "hypre"
+    pc_hypre_type = "boomeramg"
+    pc_view = True
```
Some of these options, like `ksp_view`, `ksp_monitor` and `pc_view`, are only intended for debugging. They should probably be turned off by default.

finsberg replied:
I added a verbose flag to turn these on and off (and they are by default off). If you set the log level to debug in the main function then the verbose flag will be on.

finsberg commented:
Thanks James! This is really useful! 

I added some small refactorings and made it optional to use the Krylov solver. I do get the warning about the the values not summing to 1.0 when solving the Laplace equation and using the Krylov solver. This is not happening when we use the regular solver. I am not really sure if we can tune some of the parameters in the Krylov solver in order to fix this? I made the old solver the default one for now because I don't know the implications of this requirement not being fulfilled.

I also added a failing test (marked with xfail), so if we manage to get this test to pass, then I think we are good :)

finsberg commented on tests/test_ldrb.py:
```diff
@@ -199,21 +190,61 @@ def test_lv_angles_beta():
-@pytest.mark.parametrize("fiber_space", fiber_spaces)
-def test_lv_regression(lv_mesh, fiber_space):
-    ldrb.dolfin_ldrb(lv_mesh, fiber_space)
+@pytest.mark.xfail(
+    raises=RuntimeError,
+    strict=True,
+    reason="Krylov solvers are not accrate enough?",
+)
+def test_krylov_laplace(lv_geometry):
+    ldrb.ldrb.laplace(
+        mesh=lv_geometry.mesh,
+        ffun=lv_geometry.ffun,
+        markers=lv_geometry.markers,
+        use_krylov_solver=True,
+        strict=True,
+    )
```
Here is the failing test

jamtrott commented:
> Thanks James! This is really useful!

I'm glad you found some of it useful!

> I added some small refactorings and made it optional to use the Krylov solver. I do get the warning about the the values not summing to 1.0 when solving the Laplace equation and using the Krylov solver. This is not happening when we use the regular solver. I am not really sure if we can tune some of the parameters in the Krylov solver in order to fix this? I made the old solver the default one for now because I don't know the implications of this requirement not being fulfilled.

This seems like a reasonable approach. The most obvious parameters to tune would be the convergence criteria. I think I just tried with a tolerance of 10^-9 for the relative residual. Maybe it's worth trying with a stricter tolerance to see if it helps?

> I also added a failing test (marked with xfail), so if we manage to get this test to pass, then I think we are good :)

Sounds great!

finsberg commented:
Reducing the tolerance to 1E-10 made the error go away. I think this is ready to be merged

jamtrott commented:
Good news that it worked with a stricter tolerance.

Great work, Henrik. Thanks!
